File: 1014_best_sightseeing_pair.cpp
class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& z) {
        int a = z[0], b = 0;
        for (int i = 1; i < z.size(); ++i) {
            b = max(b, a + z[i] - i);
            a = max(a, z[i] + i);
        }
        return b;
    }
};

File: 1014_best_sightseeing_pair.java
class Solution {
    public int maxScoreSightseeingPair(int[] a) {
        int maxSum = 0, maxI = a[0];
        for (int j = 1; j < a.length; j++) {
            maxSum = Math.max(maxSum, maxI + a[j] - j);
            maxI = Math.max(maxI, a[j] + j);
        }
        return maxSum;
    }
}
